### DISCUSSION TITLE
How does/will kitty fully replace tmux and other multiplexers?

### DISCUSSION THREAD

**Discussion by tmccombs** [kovidgoyal/kitty#4048] [General]
From the FAQ and github issues it is clear that @kovidgoyal  wants kitty to replace tmux/screen/any other multiplexers. 

Although the [FAQ](https://sw.kovidgoyal.net/kitty/faq/#i-am-using-tmux-and-have-a-problem) says that kitty "contains features that do all of what tmux does, but better, with the exception of remote persistence", there still a couple things that I at least don't know how to replace tmux with kitty with. 

# Local persistence

It looks like #391 would fix this. Probably. But it was closed without any apparent solution.  I'm not sure if it might still be implemented, and the thread was closed because the comments were too argumentative, or if @kovidgoyal just lost all interest in doing this. 

I'd also like to add that in addition to the reasons listed in that thread for wanting local persistence (accidentally closing a window, and switching between local and remote), you may want it if you use a multi-user computer and want to persist sessions across logins, or if you want to protect against your DE crashing (it's happened to me). 

You could use something like dtach or abduco for this, but those have some of the same problems as using tmux. And neither seems to be very actively maintained, abduco isn't in the debian/ubuntu repos, and dtach isn't in the official arch repos. There is also disown/nohup, but afiak, there isn't a way to re-attach to a process after disowning it.

# Remote multiplexing

This is where you connect to a remote host, then want to have multiple shells (or other terminal programs) simultaneously. With tmux/screen/etc. this works even if you are multiple levels of ssh, sudo/doas/su, etc. 

I know of some workarounds for this that aren't really kitty specific, and I certainly wouldn't call them "better":

- Open multiple ssh connections in separate windows. Requires multiple connections, possibly requiring entering password or 2fa credentials multiple times. Requires several steps if there are levels of nesting. May not work if the connection only supports a single connection at a time (serial port maybe? I'm not very familiar with those)
- ssh in multiple windows, but use the ControlMaster ssh option to multiplex connections. Now you only have a single connection, and only have to authenticate once, but you still have to re-enter the ssh command, and go through each step if you have nested connections or commands. Also can't be done retroactively, so you need to always enable it or know beforehand if you need it. Specific to ssh.

A kitty daemon, as mentioned in #391 could potentially allow kitty to do this natively. But the kitty daemon would need to be installed on the remote host. In many cases that's not a problem. But the user may not have permissions to install new software, or it may just be undesirable to install a new package, especially if you are the only user on the server that uses kitty. It wold also often be undesirable to pull in dependencies like X, wayland and fontconfig, so it should be possible to install the daemon separately. 

# Remote persistence

This is acknowledged in the FAQ. Similar workaround to local persistence, but dtach and abduco are less likely to be installed than screen and tmux, and as mentioned in "Remote multiplexing", installing additional software can be problematic in a remote setting. 

Am I missing ways that kitty solves any of these problems? If not, are there plans for kitty to solve these problems?

**Comment by kovidgoyal** [comment]
On Wed, Sep 22, 2021 at 02:04:12AM -0700, Thayne McCombs wrote:
> It looks like #391 would fix this. Probably. But it was closed without any apparent solution.  I'm not sure if it might still be implemented, and the thread was closed because the comments were too argumentative, or if @kovidgoyal just lost all interest in doing this. 

I have nothing against it, but I have more important/interesting/useful
things to do. And I'm afraid I am not discussing this topic with anyone
that hasn't committed to actually work on it.
